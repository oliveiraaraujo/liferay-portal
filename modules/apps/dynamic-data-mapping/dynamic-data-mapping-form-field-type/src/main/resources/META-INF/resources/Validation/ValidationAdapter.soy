{namespace ValidationAdapter}

/**
 * ValidationAdapter
 */
{template .render}
	<div key="ddm-field-type-validation"></div>
{/template}