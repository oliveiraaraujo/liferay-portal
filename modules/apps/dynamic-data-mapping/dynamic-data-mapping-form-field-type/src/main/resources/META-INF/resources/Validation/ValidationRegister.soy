{namespace ValidationRegister}

{deltemplate PageRenderer.RegisterFieldType variant="'validation'"}
	{call ValidationAdapter.render data="all" /}
{/deltemplate}