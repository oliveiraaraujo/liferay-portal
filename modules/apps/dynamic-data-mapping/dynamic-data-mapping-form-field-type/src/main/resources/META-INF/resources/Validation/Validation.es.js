/**
 * Copyright (c) 2000-present Liferay, Inc. All rights reserved.
 *
 * This library is free software; you can redistribute it and/or modify it under
 * the terms of the GNU Lesser General Public License as published by the Free
 * Software Foundation; either version 2.1 of the License, or (at your option)
 * any later version.
 *
 * This library is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 * FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License for more
 * details.
 */

import './ValidationRegister.soy';

import ClayForm from '@clayui/form';
import React, {useState} from 'react';

import {Main as Checkbox} from '../Checkbox/Checkbox.es';
import {Main as Numeric} from '../Numeric/Numeric.es';
import {Main as Select} from '../Select/Select.es';
import {Main as Text} from '../Text/Text.es';
import getConnectedReactComponentAdapter from '../util/ReactComponentAdapter.es';
import {connectStore} from '../util/connectStore.es';
import {subWords} from '../util/strings.es';
import templates from './ValidationAdapter.soy';
import {getSelectedValidation, transformData} from './transform.es';

const Validation = ({
	dataType,
	editingLanguageId,
	enableValidation: initialEnableValidation,
	errorMessage: initialErrorMessage,
	label,
	localizationMode,
	name,
	onChange,
	parameter: initialParameter,
	parameterMessage,
	readOnly,
	selectedValidation: initialSelectedValidation,
	spritemap,
	validation,
	validations,
	value,
	...otherProps
}) => {
	const [
		{enableValidation, errorMessage, parameter, selectedValidation},
		setState,
	] = useState({
		enableValidation: initialEnableValidation,
		errorMessage: initialErrorMessage,
		parameter: initialParameter,
		selectedValidation: initialSelectedValidation,
	});

	const DynamicComponent =
		selectedValidation &&
		selectedValidation.parameterMessage &&
		dataType === 'string'
			? Text
			: Numeric;

	const handleChange = (key, newValue) => {
		setState(prevState => {
			const newState = {
				...prevState,
				[key]: newValue,
			};

			let expression = {};

			if (newState.enableValidation) {
				expression = {
					name: newState.selectedValidation.name,
					value: subWords(newState.selectedValidation.template, {
						name: validation.fieldName,
					}),
				};
			}

			onChange({
				enableValidation: newState.enableValidation,
				errorMessage: {
					...value.errorMessage,
					[editingLanguageId]: newState.errorMessage,
				},
				expression,
				parameter: {
					...value.parameter,
					[editingLanguageId]: !value.expression
						? parameterMessage
						: newState.parameter,
				},
			});

			return newState;
		});
	};

	const transformSelectedValidation = getSelectedValidation(validations);

	return (
		<ClayForm.Group className="lfr-ddm-form-field-validation">
			<Checkbox
				{...otherProps}
				disabled={readOnly}
				label={label}
				name="enableValidation"
				onChange={(event, value) =>
					handleChange('enableValidation', value)
				}
				showAsSwitcher
				spritemap={spritemap}
				value={enableValidation}
				visible
			/>

			{enableValidation && (
				<>
					<Select
						{...otherProps}
						disableEmptyOption
						label={Liferay.Language.get('if-input')}
						name="selectedValidation"
						onChange={({value}) =>
							handleChange(
								'selectedValidation',
								transformSelectedValidation(value)
							)
						}
						options={validations}
						placeholder={Liferay.Language.get('choose-an-option')}
						readOnly={readOnly || localizationMode}
						spritemap={spritemap}
						value={[selectedValidation.name]}
						visible
					/>
					{selectedValidation.parameterMessage && (
						<DynamicComponent
							{...otherProps}
							dataType={dataType}
							label={Liferay.Language.get('the-value')}
							name={`${name}_parameter`}
							onChange={event =>
								handleChange('parameter', event.target.value)
							}
							placeholder={selectedValidation.parameterMessage}
							readOnly={readOnly}
							required={false}
							spritemap={spritemap}
							value={parameter}
							visible
						/>
					)}
					<Text
						{...otherProps}
						label={Liferay.Language.get('show-error-message')}
						name={`${name}_errorMessage`}
						onChange={event =>
							handleChange('errorMessage', event.target.value)
						}
						placeholder={Liferay.Language.get('show-error-message')}
						readOnly={readOnly}
						required={false}
						spritemap={spritemap}
						value={errorMessage}
						visible
					/>
				</>
			)}
		</ClayForm.Group>
	);
};

const ValidationProxy = connectStore(
	({
		emit,
		dataType: initialDataType,
		defaultLanguageId,
		dispatch,
		editingLanguageId,
		label,
		name,
		readOnly,
		spritemap,
		store,
		validation,
		validations: initialValidations,
		value = {},
	}) => {
		const data = transformData({
			defaultLanguageId,
			editingLanguageId,
			initialDataType,
			initialValidations,
			validation,
			value,
		});

		return (
			<Validation
				{...data}
				dispatch={dispatch}
				editingLanguageId={editingLanguageId}
				label={label}
				name={name}
				onChange={value => emit('fieldEdited', {}, value)}
				readOnly={readOnly}
				spritemap={spritemap}
				store={store}
				validation={validation}
				value={value}
			/>
		);
	}
);

const ReactValidationAdapter = getConnectedReactComponentAdapter(
	ValidationProxy,
	templates
);

export {ReactValidationAdapter};
export default ReactValidationAdapter;
