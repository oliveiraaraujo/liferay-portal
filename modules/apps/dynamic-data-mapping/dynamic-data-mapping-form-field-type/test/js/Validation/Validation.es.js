/**
 * Copyright (c) 2000-present Liferay, Inc. All rights reserved.
 *
 * This library is free software; you can redistribute it and/or modify it under
 * the terms of the GNU Lesser General Public License as published by the Free
 * Software Foundation; either version 2.1 of the License, or (at your option)
 * any later version.
 *
 * This library is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 * FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License for more
 * details.
 */

import {act, fireEvent} from '@testing-library/react';

import Validation from '../../../src/main/resources/META-INF/resources/Validation/Validation.es';
import withContextMock from '../__mocks__/withContextMock.es';

let component;
const spritemap = 'icons.svg';
const defaultValue = {
	errorMessage: {},
	expression: {},
	parameter: {},
};

const ValidationWithContextMock = withContextMock(Validation);

describe('Validation', () => {
	afterEach(() => {
		if (component) {
			component.dispose();
		}
	});

	it('renders checkbox to enable Validation', () => {
		component = new ValidationWithContextMock({
			dataType: 'string',
			label: 'Validator',
			name: 'validation',
			spritemap,
			value: defaultValue,
		});

		expect(component).toMatchSnapshot();
	});

	it('enables validation after click on toogle', done => {
		jest.useFakeTimers();

		const handleFieldEdited = data => {
			expect(data.value).toEqual({
				enableValidation: true,
				errorMessage: {
					en_US: undefined,
				},
				expression: {
					name: 'notContains',
					value: 'NOT(contains(textfield, "{parameter}"))',
				},
				parameter: {
					en_US: undefined,
				},
			});
			done();
		};

		const events = {fieldEdited: handleFieldEdited};

		component = new ValidationWithContextMock({
			defaultLanguageId: 'en_US',
			editingLanguageId: 'en_US',
			events,
			expression: {},
			label: 'Validator',
			name: 'validation',
			spritemap,
			validation: {
				dataType: 'string',
				fieldName: 'textfield',
			},
			value: defaultValue,
		});

		const inputCheckbox = component.element.querySelector(
			'input[type="checkbox"]'
		);

		fireEvent.click(inputCheckbox);

		act(() => {
			jest.runAllTimers();
		});
	});

	it('renders parameter field with Numeric element', done => {
		jest.useFakeTimers();

		const handleFieldEdited = data => {
			expect(data.value).toEqual({
				enableValidation: true,
				errorMessage: {
					en_US: undefined,
				},
				expression: {
					name: 'lt',
					value: 'numericfield<{parameter}',
				},
				parameter: {
					en_US: undefined,
				},
			});
			done();
		};

		const events = {fieldEdited: handleFieldEdited};

		component = new ValidationWithContextMock({
			dataType: 'numeric',
			defaultLanguageId: 'en_US',
			editingLanguageId: 'en_US',
			events,
			expression: {},
			label: 'Validator',
			name: 'validation',
			spritemap,
			validation: {
				dataType: 'integer',
				fieldName: 'numericfield',
			},
			value: defaultValue,
		});

		const inputCheckbox = component.element.querySelector(
			'input[type="checkbox"]'
		);

		fireEvent.click(inputCheckbox);

		act(() => {
			jest.runAllTimers();
		});
	});
});
