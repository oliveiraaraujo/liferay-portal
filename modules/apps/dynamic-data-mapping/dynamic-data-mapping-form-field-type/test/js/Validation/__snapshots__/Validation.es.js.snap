// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Validation renders checkbox to enable Validation 1`] = `
<div class="react-component-adapter">
  <div class="form-group lfr-ddm-form-field-validation">
    <div class="form-group" data-field-name="enableValidation"><label class="ddm-toggle-switch toggle-switch"><input class="toggle-switch-check" name="enableValidation" type="checkbox" value="true"><span aria-hidden="true" class="toggle-switch-bar"><span class="toggle-switch-handle"></span><span class="toggle-switch-text toggle-switch-text-right">Validator</span></span></label></div>
  </div>
</div>
`;
